RecoHGCal/GraphReco/plugins/peprCandidateFromHitProducer.cc

```diff
@@ -0,0 +1,345 @@
+
+    // rechit tools
+    hgcal::RecHitTools recHitTools_;
+
+    // windows
+    std::vector<InferenceWindow> windows_;
+
+    double minEta_;
+    double maxEta_;
+    double etaFrameWidth_;
+    double phiFrameWidth_;
+    size_t nEtaSegments_;
+    size_t nPhiSegments_;
+
+
+};
+
+
+peprCandidateFromHitProducer::peprCandidateFromHitProducer(const edm::ParameterSet& config) :
+        recHitCollections_(config.getParameter<std::vector<edm::InputTag> >("recHitCollections")), 
+        tracksToken_(consumes<edm::View<reco::Track>>(config.getParameter<edm::InputTag>("tracks"))),
+        simClusterToken_(consumes<std::vector<SimCluster>>(config.getParameter<edm::InputTag>("simClusters"))), 
+        batchedModel_(config.getParameter<bool>("batchedModel")), 
+        tritonPath_(config.getParameter<std::string>("tritonPath")), 
+        inpipeName_(config.getParameter<std::string>("inpipeName")),
+        outpipeName_(config.getParameter<std::string>("outpipeName")),
+        //FIXME: actually these are all not needed if windows are created in the constructor!
+        minEta_(config.getParameter<double>("minEta")),
+        maxEta_(config.getParameter<double>("maxEta")),
+        etaFrameWidth_(config.getParameter<double>("etaFrameWidth")),
+        phiFrameWidth_(config.getParameter<double>("phiFrameWidth")),
+        nEtaSegments_((size_t)config.getParameter<uint32_t>("nEtaSegments")),
+        nPhiSegments_((size_t)config.getParameter<uint32_t>("nPhiSegments"))
+        {
+
+    // sanity checks for sliding windows
+
+    // get tokens
+    for (edm::InputTag& recHitCollection : recHitCollections_) {
+        recHitTokens_.push_back(consumes<HGCRecHitCollection>(recHitCollection));
+    }
+   
+
+    produces<reco::PFCandidateCollection>();
+    //produces<std::vector<Trackster>>();
+
+
+    //launch scripts to work with Triton clients
+    //https://github.com/cms-pepr/HGCalML/tree/master/triton
+    
+    std::string str0 = "cd " + tritonPath_;
+    const char *cdcommand = str0.c_str();
+    system(cdcommand);
+
+    std::string str1 = "./cmssw_oc_server.sh > serverlog.txt &";
+    const char *clientcommand1 = str1.c_str();
+    std::cout << "Executing: " << clientcommand1 << std::endl;
+    system(clientcommand1); 
+
+    std::cout << "Sleeping 30s... " << std::endl;
+    system("sleep 30");
+
+    std::string str2 = "./cmssw_oc_forward_client.sh " + inpipeName_;
+    const char *clientcommand2 = str2.c_str();
+    std::cout << "Executing: " << clientcommand2 << std::endl;
+    system(clientcommand2);
+
+    system("cd -");
+
+}
+
+peprCandidateFromHitProducer::~peprCandidateFromHitProducer() {
+
+    //FIXME: find process ID of scripts run in the constructor
+    //system("killtask <...> 15");
+    //system("killtask <...> 15");
+}
+
+
+void peprCandidateFromHitProducer::beginStream(edm::StreamID streamId) {
+    windows_ = InferenceWindow::createWindows(nPhiSegments_,nEtaSegments_,minEta_,maxEta_,etaFrameWidth_,phiFrameWidth_);
```

Comment:
This function will automatically take care of creating one window for positive and one for negative eta.

Reply:
I think it works although I am not 100% sure yet if the streams for different windows are not stacking up.

Reply:
what do you mean by "stacking up"?

Reply:
In a local test 'appending' to an existing file (as test for appending to an existing pipe) results in the input arrays of the next window to be added to the same file as the arrays of the first window. I was just not sure if with actual pipes this is not a problem; I think not because the model is read after the first window, and when switching to the second window the input of the first is not considered anymore.

Reply:
I mean, you write and read the output in a loop as far as I can see.
So the next window will only be processed if the first is done. So there shouldn't be any overlap

Reply:
Yes, you are right (I just couldn't check yet if it was behaving as it should with a real model giving actual output).